Increase internal algorithm security using HS512 and 128 byte hmac keys

rmartinc opened the pull request:
Closes #13080

PR that increase security in default internal algorithm (used by actions, refresh tokens,...). As discussed in the issue the default hmac size has been updated from 64 to 128 bytes and the internal algorithm has changed from HS256 to HS512. Things to consider:

1. We can maintain default hmac size (64) and just increase the size for the initial provider. If you prefer this solution just let me know.
2. The new HS512 internal key is called `hmac-generated-hs512` instead of previous `hmac-generated`. I decided to do that to differentiate the provider names in migrations. But maybe we prefer to maintain the same name or rename old HS256 for compatibility.

It's just a draft for now. We also need to decide if we want this in keycloak 24 or we wait for 25. For the moment the upgrade task is for 24 but it can be moved to 25. An upgrading note commenting the final changes is also needed before moving out from draft.

rmartinc committed:
Increase internal algorithm security using HS512 and 128 byte hmac keys
Closes #13080

Signed-off-by: rmartinc <rmartinc@redhat.com>

mposolda reviewed (commented):
- Added one nitpick comment.

- Besides that, it seems that there are some conflicts in the meantime.

- Did some search for the `Algorithm.HS256` and found few more places using that as a default. Wonder if we should try to update those as well when we are at this? Some of the places I found (might missed some):
  - `Attributes.HS_ALGORITHM_PROPERTY` (This probably causes that HS256 is still pre-selected as the default algorithm in the UI for the configuration properties of some providers)
  - `AbstractOAuth2IdentityProvider.getSignatureContext()` - looks this one is also documented here https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker_oidc (see "client assertion signature algorithm" section), so might change the docs too
  - `CIBAAuthenticationRequest` - but here I am not 100% sure if CIBA specification explicitly mentions HS256 should be the default, so might be good to doublecheck...

- Do we also want to mention this a bit in `changes-24_0_0.adoc` ? It is internal in theory, but still can be nice to mention this, so people are aware

Finally, it may be probably good to add some minor update to

rmartinc commented on services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java:
```diff
@@ -63,6 +63,7 @@
 
 public class DefaultTokenManager implements TokenManager {
 
+    public static final String INTERNAL_ALGORITHM = Algorithm.HS512;
```
OK, Done!

rmartinc commented:
>       * `Attributes.HS_ALGORITHM_PROPERTY` (This probably causes that HS256 is still pre-selected as the default algorithm in the UI for the configuration properties of some providers)

I have changed to set HS512 as the default one now.

>       * `AbstractOAuth2IdentityProvider.getSignatureContext()` - looks this one is also documented here https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker_oidc (see "client assertion signature algorithm" section), so might change the docs too

I have not done this point. Changing this can break current configurations and this is really not related to internal keycloak signature. This is just using the client secret to create a JWT for the broker authentication, any other HS algorithm can be configured and it's an admin decision. If you think it's important I will do it but take in mind that configurations with no alg specified are changing from HS256 to HS512 and can break.

>       * `CIBAAuthenticationRequest` - but here I am not 100% sure if CIBA specification explicitly mentions HS256 should be the default, so might be good to doublecheck...

Done! I have checked that this direct signature is used to create the `auth_req_id` which, in the spec, is just an ID. The problem with this change is that previous IDs will be invalid after migration (as the token is using direct JWE enc/sig) and a new authentication is needed. But this was thought with no idea of rating keys, so I just changed it anyway.

>     * Do we also want to mention this a bit in `changes-24_0_0.adoc` ? It is internal in theory, but still can be nice to mention this, so people are aware

Yep, for sure. I'll do when the other comments are OK.

mposolda commented:
@rmartinc Thanks for the updates! So perhaps only migration docs and then we're good to go?

rmartinc commented:
Done! I added a little upgrading section explaining we have changed from HS256 to HS512 for internal tokens. If tests are OK tomorrow will move it out of draft.

mposolda commented:
@andymunro Do you please have a chance to review the documentation part of this PR?

ahus1 commented:
@kami619 - this change might (will?) affect Keycloak's performance.

Once this is merged and available in a nightly release, please keep an eye on the nightly KCB runs if we need to update our sizing guide. Please add a new issue to the Keycloak project so we can track this, and update that issues's milestone for KC24 or KC25 depending on when this is being merged. Thanks!

kami619 commented:
@ahus1 thanks, I will track this.

mposolda commented:
@ahus1 @kami619 This is good point. Sorry we did not explicitly pinged you about this.
